src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.SourceGenerators.Testing/SourceGeneratorTest`1.cs

```diff
@@ -4,14 +4,139 @@
+                Verify.NotEmpty($"{nameof(TestState)}.{nameof(SolutionState.Sources)}", TestState.Sources);
+            }
+
+            var analyzers = GetDiagnosticAnalyzers().ToArray();
+            var defaultDiagnostic = GetDefaultDiagnostic(analyzers);
+            var supportedDiagnostics = analyzers.SelectMany(analyzer => analyzer.SupportedDiagnostics).ToImmutableArray();
+            var fixableDiagnostics = ImmutableArray<string>.Empty;
+            var rawTestState = TestState.WithInheritedValuesApplied(null, fixableDiagnostics);
+            var rawIncrementalState = IncrementalChangeTestState.WithInheritedValuesApplied(null, fixableDiagnostics);
+
+            var testState = rawTestState.WithProcessedMarkup(MarkupOptions, defaultDiagnostic, supportedDiagnostics, fixableDiagnostics, DefaultFilePath);
+            var incrementalTestState = rawIncrementalState.WithProcessedMarkup(MarkupOptions, defaultDiagnostic, supportedDiagnostics, fixableDiagnostics, DefaultFilePath);
+
+            var diagnostics = await VerifySourceGeneratorAsync(testState, incrementalTestState, Verify, cancellationToken).ConfigureAwait(false);
+            await VerifyDiagnosticsAsync(new EvaluatedProjectState(testState, ReferenceAssemblies), testState.AdditionalProjects.Values.Select(additionalProject => new EvaluatedProjectState(additionalProject, ReferenceAssemblies)).ToImmutableArray(), testState.ExpectedDiagnostics.ToArray(), Verify, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> VerifySourceGeneratorAsync(SolutionState testState, SolutionState incrementalTestState, TVerifier verifier, CancellationToken cancellationToken)
+        {
+            var sourceGenerators = GetSourceGenerators().ToImmutableArray();
+            if (sourceGenerators.IsEmpty)
+            {
+                return ImmutableArray<Diagnostic>.Empty;
+            }
+
+            var driver = await VerifySourceGeneratorAsync(Language, sourceGenerators, testState, verifier.PushContext("Source generator application"), cancellationToken);
+
+            // If any solution transforms for incremental generator testing have been defined,
+            // then test incremental generators.
+            // We'll also only go down here if incremental generators exist.
+            if (incrementalTestState is not null && typeof(CompilationOptions).Assembly.GetType("Microsoft.CodeAnalysis.IIncrementalGenerator") is not null)
+            {
+                var secondRunProject = await CreateProjectAsync(new EvaluatedProjectState(incrementalTestState, ReferenceAssemblies), incrementalTestState.AdditionalProjects.Values.Select(additionalProject => new EvaluatedProjectState(additionalProject, ReferenceAssemblies)).ToImmutableArray(), cancellationToken);
+
+                driver = driver
+                        .ReplaceAdditionalTexts(secondRunProject.AnalyzerOptions.AdditionalFiles)
```

Comment:
❔ Will the driver recognize cases where a file did not change? For example, consider the case of two additional files. If you change the content of just one of them, will it treat the other one as unchanged?